runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/SdkHarnessClient.java

```diff
@@ -476,6 +511,38 @@ public void abort() {}
     }
   }
 
+  /**
+   * A {@link CloseableFnDataReceiver} which counts the number of elements that have been accepted.
+   */
+  private static class CountingFnDataReceiver<T> implements CloseableFnDataReceiver<T> {
+    private final CloseableFnDataReceiver delegate;
+    private long count;
+
+    private CountingFnDataReceiver(CloseableFnDataReceiver delegate) {
+      this.delegate = delegate;
+    }
+
+    public long getCount() {
+      return count;
+    }
+
+    @Override
+    public void accept(T input) throws Exception {
+      count += 1;
```

Comment:
I'm not sure whether the data channel does something like buffering. I'm curious whether the count should be counted `when `flush` is called?

Reply:
In this case the runner has committed to sending the value to the SDK, we can't realistically account for all "buffers" that may live between this call and when the SDK receives it (JVM buffers, OS buffers, network buffers, ...).

Reply:
Thanks! Just wanna make sure we don't do any manual buffer in between.